Make parse error suggestions verbose and fix spans

estebank opened the pull request:
Go over all structured parser suggestions and make them verbose style.

When suggesting to add or remove delimiters, turn them into multiple suggestion parts.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

estebank committed:
Make parse error suggestions verbose and fix spans

Go over all structured parser suggestions and make them verbose style.

When suggesting to add or remove delimiters, turn them into multiple suggestion parts.

oli-obk commented:
> Go over all structured parser suggestions and make them verbose style.

why? Most of these seemed fine to me



> When suggesting to add or remove delimiters, turn them into multiple suggestion parts.

For these It makes sense

oli-obk commented on tests/ui/async-await/await-keyword/incorrect-syntax-suggestions.stderr:
```diff
@@ -2,115 +2,214 @@ error: incorrect use of `await`
   --> $DIR/incorrect-syntax-suggestions.rs:8:13
    |
 LL |     let _ = await bar();
-   |             ^^^^^^^^^^^ help: `await` is a postfix operation: `bar().await`
+   |             ^^^^^^^^^^^
+   |
+help: `await` is a postfix operation
+   |
+LL |     let _ = bar().await;
+   |             ~~~~~~~~~~~
```
I guess it would make sense to make this a verbose suggestion if the `await ` removal were one span and the `.await` addition another.

But the way it is currently it just seems redundant.

oli-obk commented on tests/ui/did_you_mean/issue-41679-tilde-bitwise-negation-attempt.stderr:
```diff
@@ -2,39 +2,56 @@ error: `~` cannot be used as a unary operator
+   |
+help: use `!` to perform logical negation or bitwise not
+   |
+LL |     let _v = !v;
+   |              ~
```
this is a bit weird, it doesn't highlight the `not ` that is going to be removed for this

estebank replied:
This is changed by #127541.

oli-obk commented on tests/ui/enum/nested-enum.stderr:
```diff
@@ -2,25 +2,37 @@ error: `enum` definition cannot be nested inside `enum`
-   |     help: consider creating a new `struct` definition instead of nesting
+   |     ^^^^^^
+   |
+help: consider creating a new `struct` definition instead of nesting
+   |
+LL -     struct Quux { field: u8 },
+LL +
```
uhm?

estebank replied:
Pre-existing. It happens when we suggest removing something but leave whitespace behind. I know I fixed this in a separate PR that might not have been merged yet, but can't find the commit 😵‍💫

estebank replied:
Oh, I just found what the problem: `//~ ERROR` is removed after the stderr file is created, the suggestion makes more sense when running against the file manually.

oli-obk commented on tests/ui/generics/issue-95208-ignore-qself.stderr:
```diff
@@ -2,9 +2,12 @@ error: expected `:` followed by trait or lifetime
-   |                                                                    help: use single colon: `:`
+   |                                                                                        ^
+   |
+help: use single colon
+   |
+LL | impl<T: Iterator> Struct<T> where <T as std:: iter::Iterator>::Item: std::fmt::Display {
```
this could be a removal suggestion instead of a replacement suggestion

estebank replied:
This one is tricky, because this could have been written as `<T as std:: iter::Iterator>::Item  ::    std::fmt::Display`, but we only have the spans for `<T as std:: iter::Iterator>::Item` and `std::fmt::Display`.

oli-obk commented on tests/ui/impl-trait/extra-impl-in-trait-impl.stderr:
```diff
@@ -2,25 +2,35 @@ error: unexpected `impl` keyword
   --> $DIR/extra-impl-in-trait-impl.rs:12:6
    |
 LL | impl impl Default for S2 {
-   |      ^^^^^ help: remove the extra `impl`
+   |      ^^^^^
    |
 note: this is parsed as an `impl Trait` type, but a trait is expected at this position
   --> $DIR/extra-impl-in-trait-impl.rs:12:6
    |
 LL | impl impl Default for S2 {
    |      ^^^^^^^^^^^^
+help: remove the extra `impl`
+   |
+LL - impl impl Default for S2 {
+LL + impl Default for S2 {
+   |
```
I think the original was better

oli-obk commented on tests/ui/macros/missing-bang-in-decl.stderr:
```diff
@@ -24,7 +24,13 @@ error: macro names aren't followed by a `!`
   --> $DIR/missing-bang-in-decl.rs:10:16
    |
 LL | macro_rules bar! {
-   |                ^ help: remove the `!`
+   |                ^
+   |
+help: remove the `!`
+   |
+LL - macro_rules bar! {
+LL + macro_rules bar {
+   |
```
same here, the verbose suggestion is kinda noisy

bors commented:
:umbrella: The latest upstream changes (presumably #127419) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank committed:
Make `impl` and `!` removal suggestion `short`

estebank committed:
Tweak tests to avoid confusing suggestion output

estebank committed:
Use more accurate span for `:` to `::` suggestion

estebank committed:
More accurate incorrect use of `await` suggestion

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9900610745/job/27351686608) [(plain)](https://github.com/rust-lang/rust/commit/e4df175a00ea28b0a76fce3ea3026bb39a45b346/checks/27351686608/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 -g
#12 5.414 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.476 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.485 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 5.536 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 5.566 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 5.574 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 6.567 
#12 6.567 added 205 packages in 6s
#12 6.567 
#12 6.567 20 packages are looking for funding
```

</details>

estebank committed:
fix unused binding

estebank committed:
Make `;` suggestions inline

oli-obk commented:
As discussed in our 1:1, this needs a motivation and some links to annotate snippets discussions

estebank commented:
The initial impetus to start this work is to facilitate [the migration to annotate-snippets](https://rust-lang.github.io/rust-project-goals/2024h2/annotate-snippets.html), which [doesn't yet support suggestions](https://github.com/rust-lang/annotate-snippets-rs/issues/89), but @Muscraft is working on. Sadly the conversations were 1:1, but the gist of them is:

 - a precondition for migrating is minimal changes to the output, we'll accept small divergences as long as they are reasonable, but we're aiming for 100% mimicry
 - annotate-snippets is unlikely to ever support "inline suggestions"
 - rustc is the one dealing with json output where these are relevant, so rustc can continue transforming suggestions into labels for annotate-snippets' sake
 - Minimizing how many of the inline suggestions exist will help with the migration
 - The inline suggestions were the first ones to be implemented; over time we supported multiple suggestions, suggestions with multiple parts, and rendering these suggestions with underlines for addition and replacements and a diff format for removals. I am of the opinion that the "verbose" renderings are easier to understand for people for any case beyond "add an `x` here" or `remove this `x`", particularly because suggestions where the code is similar in both cases it is easier to *see* the difference when presented side by side (people are already used to the patch format that diff tools use), color further aids with this. For TTS users, none of the available outputs are *good*.
 - Making suggestions verbose is dredging up several suggestions that were incorrectly written (bad spans, using them as span_help, bad code being suggested) that were harder to spot with the inline suggestions; there are also many cases where the suggestion message is something non-descriptive like "try:", which also needs to be changed (editors can show the message without the code, so the message should be enough for someone that understands what's going on to fix their own code)

The additional commits after the one you initially reviewed are addressing your review comments.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 71f16bdb32feb33762e5245efa37edac2bde724b has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 71f16bdb32feb33762e5245efa37edac2bde724b -->
<!-- homu: {"type":"Approved","sha":"71f16bdb32feb33762e5245efa37edac2bde724b","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

estebank commented on tests/ui/parser/bad-char-literals.stderr:
```diff
@@ -2,27 +2,47 @@ error: character constant must be escaped: `'`
-   |      ^ help: escape the character: `\r`
+   |      ^
+   |
+help: escape the character
+   |
+LL |     '\r';
+   |      ++
```
I'm noticing that the *renderer* considers *any* whitespace as "not there" for the purposes of deciding whether this is a replacement or addition (this is normally done to have better output), but it has the side effect of rendering not quite correctly in the face of suggestions to replace zero-width whitespace characters.
